### ISSUE TITLE
incorrect/missing maven settings file (for plugin builds?)

### ISSUE THREAD

**Issue by jtnord** [jenkins-infra/helpdesk#3101]
### Service(s)

ci.jenkins.io

### Summary

https://github.com/jenkinsci/google-metadata-plugin/pull/22/checks failed because

```
[2022-08-19T11:47:32.752Z] [ERROR] [ERROR] Some problems were encountered while processing the POMs:
[2022-08-19T11:47:32.752Z] [FATAL] Non-resolvable parent POM for org.jenkins-ci.plugins:google-metadata-plugin:${revision}${changelist}: Could not find artifact org.jenkins-ci.plugins:plugin:pom:4.46 in central (https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 20, column 11
[2022-08-19T11:47:32.752Z]  @ 
[2022-08-19T11:47:32.752Z] [ERROR] The build could not read 1 project -> [Help 1]
[2022-08-19T11:47:32.752Z] org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[2022-08-19T11:47:32.752Z] [FATAL] Non-resolvable parent POM for org.jenkins-ci.plugins:google-metadata-plugin:${revision}${changelist}: Could not find artifact org.jenkins-ci.plugins:plugin:pom:4.46 in central (https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 20, column 11
(https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 20, column 11
```

This means that the maven settings files in use for the plugin builds (why would they be different for linux agents) are not forcing all artefact retrieval to go via `repo.jenkins.io/public`   Plugins etc should not be able to add new repos to download artifacts from outside of our repository.  
This is partially to ensure that we are only hosting/building open source plugins, partially to ensure we cache all releases we download, and also for security reasons.

In other words the plugin build should be using a maven settings file with `mirrorof` `*,!incrementals` that points to public and another mirror that mirrors the incrementals repos.

### Reproduction steps

1. not sure - a standard plugin build where the plugin does not define any repositories?

check the logs at https://github.com/jenkinsci/google-metadata-plugin/pull/22/checks

**Comment by dduportal** [comment]
I'm not sure why does it fail specifically for Windows agents, but I confirm that it is not expected, by the jenkins infra, that a custom settings.xml file is used on any agent for ci.jenkins.io.

I assume it means that all pom.xml of each plugin is expected to specify their repositories (I assume repo.jenkins.io at least?), as per https://maven.apache.org/guides/mini/guide-mirror-settings.html#using-mirrors-for-repositories 

> Repositories can be [declared inside a project](https://maven.apache.org/guides/mini/guide-multiple-repositories.html), which means that if you have your own custom repositories, those sharing your project easily get the right settings out of the box.

Reminder that repo.jenkins.io is not expected to be a cache of "everything" given the size and bandwidth involved.

However we are working on a caching proxy, to decrease the risk of going crazy for ci.jenkins.io in #2752 .

**Comment by dduportal** [comment]
@jtnord about the PR you shared:

-  Build #1 (commit 3f5ed4117f0c7e4d98c563f1caf88479f16e242e ) failed after ~20s in windows, and the linux build was cancelled: I can't find any log output showing that Linux successfully ran the mvn steps
- Build #2 failed for Linux, so it does not seem to have a difference between operating system
- Build #3 succeeded thanks to your commit adding the repositories: https://github.com/jenkinsci/google-metadata-plugin/pull/22/commits/d4187c9bcec98beb6ee12631ca121aea86c0e2f0 

Did I miss something?

**Comment by jtnord** [comment]
sorry I thought I linked to the specific check with the failure -> https://github.com/jenkinsci/google-metadata-plugin/pull/22/checks?check_run_id=7917477338

> Reminder that repo.jenkins.io is not expected to be a cache of "everything" given the size and bandwidth involved.

Are you sure, I would double check that with @daniel-beck - we keep all incrementals just for fun so actual releases should not be any issues :)

Possibly you could exclude `central` as that is a well behaving repo - but stull.

**Comment by dduportal** [comment]
There are a lot of elements here, let me try to list them and answer.

But I fee like this issue should be closed as it is not clear was is expected to be done by the jenkins infra: if you disagree for closing, can you elaborate on the expectation please?

- Repositories described per project's `pom.xml` that is not for the infra-team to decide. Maven documentation recommands to use explicit repositories (https://maven.apache.org/guides/mini/guide-mirror-settings.html#using-mirrors-for-repositories) to not rely only on mirrors, but it seems to be a tip more than a recommandation
- Maven settings file: There has not been any `settings.xml` file on ci.jenkins.io since 2020. 
- Herve is working on adding some in order to use a "mirror caching proxy" (that would be different services: 1 per cloud).
- I feel like we can improve the safety of our build processes by working on this topic. But I would focus only on the maven settings (if there are any: I have not check yet) for the Jenkins Core releases  (release.ci.jenkins.io) and update center generation (trusted.ci.jenkins.io).
  - Why not for ci.jenkins.io?
    - It would add a LOT of bandwidth download to the "mirror" given the amout of builds. JFrog's repo.jenkins-ci.org is already using way too much of bandwidth
    - Risks are limited because ci.jenkins.io does not deploy or release anything: it's aimed only to build and test.


Looks like your questions are related to #2322 . Do you agree?

**Comment by jtnord** [comment]
> It would add a LOT of bandwidth download to the "mirror" given the amout of builds. JFrog's repo.jenkins-ci.org is already using way too much of bandwidth

I think you will be surprised.  Given the plugins define this repo - and any others this causes a lookup to be sent and then a 404 to be returned or the data.
https://maven.apache.org/guides/mini/guide-multiple-repositories.html#repository-order

As there is no global nor local settings the order used for resolution will be the order of the repos defined in the project.
As central is not defined at the project level you will probably find are hitting the proxy anyway even if you don't expect it.

Additionally - 

if the plugin is inceementallyfied you may hit the incremental repo **and** the normal repo which is extra lookup (a mirror of incrementals could contain public and incrementals potentially reducing the lookups and then overall HTTP requests should the pom ordering be correct)

> Risks are limited because ci.jenkins.io does not deploy or release anything: it's aimed only to build and test.

but what about CD?  if you are saying that CD is not a release then that is wrong.
if you are saying CD builds should use different mirror setup to CI then I would argue that is also bad as you may get a CD build that fails where a CI build passed as the repositories are different.

**Comment by dduportal** [comment]
> > It would add a LOT of bandwidth download to the "mirror" given the amout of builds. JFrog's repo.jenkins-ci.org is already using way too much of bandwidth
> 
> I think you will be surprised. Given the plugins define this repo - and any others this causes a lookup to be sent and then a 404 to be returned or the data. https://maven.apache.org/guides/mini/guide-multiple-repositories.html#repository-order
> 
> As there is no global nor local settings the order used for resolution will be the order of the repos defined in the project. As central is not defined at the project level you will probably find are hitting the proxy anyway even if you don't expect it.
> 
> Additionally -
> 
> if the plugin is inceementallyfied you may hit the incremental repo **and** the normal repo which is extra lookup (a mirror of incrementals could contain public and incrementals potentially reducing the lookups and then overall HTTP requests should the pom ordering be correct)

Sorry but I do not understand at all. We are trying to decrease the amount of bandwitdh from repo.jenkins. I'm not sure to understand the link with the Maven's internal repository order. Can you clarify please?

If i understand correctly, your request is to add a settings.xml on all agents of ci.jenkins.io to always use repo.jenkins-ci.org as a mirror for all repositories, wether they are declared in `pom.xml` or not?
Or are you asking something else (and then I need you to clarify what are you asking from the infra team)?


> 
> > Risks are limited because ci.jenkins.io does not deploy or release anything: it's aimed only to build and test.
> 
> but what about CD? if you are saying that CD is not a release then that is wrong. if you are saying CD builds should use different mirror setup to CI then I would argue that is also bad as you may get a CD build that fails where a CI build passed as the repositories are different.

Good point. Though CD is not executed in a Jenkins controller but in a GitHub action. Worth checking its settings.

**Comment by daniel-beck** [comment]
> Reminder that repo.jenkins.io is not expected to be a cache of "everything" given the size and bandwidth involved.

repo.j.o provides mirrors ("remote" repositories) of the most common repositories, which are populated on demand. Most Jenkins plugins, at least the "important" ones, get all their dependencies from there. This is part of the archetype poms, e.g. https://github.com/jenkinsci/archetypes/blob/0f7c7446c7b258390df573fcbf6caa6051527b9a/empty-plugin/src/main/resources/archetype-resources/pom.xml#L50-L61

However, some plugins define other repositories in the pom.xml, and there is no project policy against that. Their builds would start failing if we forced repo.j.o/public to be a mirror of everything.

**Comment by dduportal** [comment]
Thanks for the insights folks! To avoid spreading the knowledge too thin, I'll close this issue as duplicate of https://github.com/jenkins-infra/helpdesk/issues/2322 + report back your comments in https://github.com/jenkins-infra/helpdesk/issues/2322. 

Reason is that the initial subject (Windows / Linux maven settings difference) was not proven and the initial issue was solved by adding the repositories in the pom.xml: there is no further action from the infra team required.

Thanks a lot!
